### ISSUE TITLE
Flat combining

### ISSUE THREAD

**Issue by schets** [aturon/crossbeam#63]
Is [flat combining](https://www.cs.bgu.ac.il/~hendlerd/papers/flat-combining.pdf) something that would fit well in crossbeam?

The jist of flat combining is that a single thread acquires ownership of a given datastructure and contending threads send messages to the owning thread which carries out the requisite tasks.

In a sense, flat-combining can be though of as:

  * A glorified mutex where instead of waiting to do work threads have the owner do work
  * A basic message-passing system where threads take over/release time as the 'worker thread'

This can also transform an spmc queue into an 'mp'mc queue, single-writer into multi-writer, etc. However, it is blocking so I'm not really sure how well this fits in with the rest of the library.

If we want to get fancy, threads can send asynchronous messages as well where they don't block.

Also, this kinda brings up a more detailed (and honestly irrelevant at this point) question of how should threads yield? In something like mio-co, an evented system, or somewhere where 'blocking' time could really just do cleanup work, spinning on a lock and calling thread::yield() is really suboptimal.

**Comment by aturon** [comment]
Absolutely; I'd love to experiment with flat combining. I'd like crossbeam to take on the full spectrum of scalable concurrent algorithms, whether non-blocking or not. (And of course, even for the non-blocking data structures, there's often support for blocking operations).

**Comment by schets** [comment]
Sounds good - I'll have a basic version ready soon and with some more thought/experimenting, something more generically usable.

**Comment by schets** [comment]
A year late I actually have time to work on this for real, and just need to port [the C implementation](https://github.com/schets/c_flat_combining). Something I've realized while working on test implementations is that flat combining has a huge advantage in that it readily supports asynchronous work (see [here](https://github.com/schets/c_flat_combining/blob/arbitrary_entry/src/combiner.h#L42)) such that work in the critical section is actually performed before the thread tries to block; it's not just a glorified try_lock loop.

There are a few ways to do flat combining that favor different use cases. Rayon's implementation of flat combining favors a situation where the same set of threads are hitting the combiner with a large amount of work. I'm not sure if this is suitable for general purpose. The implementation I linked works great for general purpose and isn't too complicated, but might not fare too well with many numa nodes competing under contention compared to something with per-node queues.

I favor going with the simple and general purpose method at first and then branching out as needed.
